### ISSUE TITLE
Allow more credential factories to be specified

### ISSUE THREAD

**Issue by joaoandremartins** [GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#41]
Currently, only [`ApplicationDefaultCredentialFactory`](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/master/core/src/main/java/com/google/cloud/sql/core/SslSocketFactory.java#L139) is used as credential factory.

It is possible to set a [`CREDENTIAL_FACTORY_PROPERTY`](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/master/core/src/main/java/com/google/cloud/sql/core/SslSocketFactory.java#L128) system property to specify another credential factory, although there is also a mention [not to use this property besides tests](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/master/core/src/main/java/com/google/cloud/sql/core/SslSocketFactory.java#L83).

In the [spring-cloud-gcp](https://github.com/spring-cloud/spring-cloud-gcp) project, it's possible for users to use credentials that aren't the application default ones (e.g., service accounts). We would like to provide more features to Spring users, such as only specifying the database and instance name (not instance connection name) and letting them use the already configured credentials is a strong requirement.

An idea is to accept another parameter in the connection string, like `credentialsFactory=com.your.connection.Factory`, similarly to `socketFactory`.

Also, in the meantime, is it OK to use the `CREDENTIAL_FACTORY_PROPERTY` property?

**Comment by laixer** [comment]
Can you clarify the problem you are trying to solve? Is it a hypothetical problem or a real problem that customers have reported?
Customers can already specify a different credential file using GOOGLE_APPLICATION_CREDENTIALS but this may not be convenient in some cases and limits you to a single override for the entire application. In practice, this doesn't seem different from setting CREDENTIAL_FACTORY_PROPERTY. 
I'm open to allowing users to specify a different credential file using an additional JDBC URL property but I'm less keen on allowing users to specify arbitrary credential factories as that could be a problem from a support perspective. Would allowing you to pass a credential file instead of credential factory be sufficient for you?

w.r.t. specifying instance name and not instance connection name: How are you planning to achieve this? The library intentionally requires the instance connection name and not the instance name.

**Comment by joaoandremartins** [comment]
A credential file in the connection string, instead of a credential factory, could work for us, but only in a specific sub-set of cases.

To tell you a bit more about our context, we're working on integrating the Spring Framework with GCP. One of Spring's objectives is to reduce boilerplate and simplify the user experience.

In the Cloud SQL specific case, our goal is to allow users to specify something like this in their Spring properties file:

```
spring.cloud.gcp.sql.instanceName=test-instance
spring.cloud.gcp.sql.databaseName=test-database
spring.cloud.gcp.sql.userName=root
spring.cloud.gcp.sql.password=pw

spring.cloud.gcp.credentialsLocation=file://usr/.../file.json
```

Our Spring Boot starter (some kind of adapter) would construct the JDBC URL and create a `DataSource` (e.g., HikariCP) that uses your `SocketFactory` and `SslSocketFactory` under the hood. It uses Cloud SQL's SqlAdmin to get the region of the instance, which is probably what you are trying to avoid doing yourselves in order to improve performance, but is a poor user experience in my opinion.

Users would then be able to just inject a [JdbcTemplate](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html) object that is pre-configured and lets them interact with their database.

One issue is that credentials are a [Spring Resource](https://docs.spring.io/spring/docs/3.2.x/spring-framework-reference/html/resources.html#resources-implementations) and, as such, can come from a multitude of places, not just the file system. So that's one considerable limitation about only specifying file locations through the connection string. In any case, every Google API I worked with provides the ability to pass in a GoogleCredential object (e.g., SQL Admin API). It seems a considerable limitation not being able to do that here.

**Comment by laixer** [comment]
Ok, I see your point about being able to support non-file credential files. I can see how it would be a poor experience if they were able to use spring resource specifiers in some places but not others. I'll make a PR to expose the credential factory as a URL parameter. 

With regards to region: the reason we require the instance connection name is because a future version of the API will likely require the region to be specified in the request. If the users don't specify the region now, this will likely mean down the road all your users will have to go through their configs and update them.

**Comment by laixer** [comment]
I remembered that the Postgres driver doesn't allow passing arbitrary properties to the socket factory so using a JDBC property may not be a good choice. It's less elegant but I think a system property is the only way this will work across both MySQL and Postgres. Do you have any concerns with that?

**Comment by joaoandremartins** [comment]
Not off the top of my head. If it's the only possible way across all drivers, so be it.

**Comment by joaoandremartins** [comment]
Any updates here? Thanks!

**Comment by laixer** [comment]
Released in 1.0.4
Configurable via cloudSql.socketFactory.credentialFactory system property.
This property name is also accessible via com.google.cloud.sql.CredentialFactory.CREDENTIAL_FACTORY_PROPERTY
